processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -407,35 +216,56 @@ public T get(int index)
   @Override
   public int indexOf(T value)
   {
-    return bufferIndexed.indexOf(value);
+    return indexOf(this, value);
+  }
+
+  private int indexOf(Indexed<T> indexed, T value)
```

Comment:
Talking about "own polymorphism", https://shipilev.net/blog/2015/black-magic-method-dispatch/ shows that even more compex forms of it (byte/int "enum", instead of boolean flag) are usually optimized better by Hotspot JIT than "real" polymorphism, because of potentially unlimited number of options and requirement to support new options in runtime (when new implementation class is loaded).